docs/test-guidelines.md

```diff
@@ -0,0 +1,57 @@
+# Test guidelines
+
+At MetaMask, we are committed to delivering high-quality software. A key aspect of achieving this goal is through a robust testing strategy that includes both end-to-end (e2e), integration and unit tests. This document provides comprehensive guidelines on each testing type, emphasizing the importance of choosing the right test for the right scenario and detailing the specific practices for writing and organizing these tests.
+
+## What are unit tests?
+
+Unit tests understanding is a bit more diffuse. As there is no consensus on how they should be written and what they should test. But generically, [as captured by Martin Fowler](https://martinfowler.com/articles/2021-test-shapes.html), there's two lines of thought: solitary unit tests and sociable unit tests. Within MetaMask there are multiple examples from tests and discussions highlighting the two types (for example you can check the discussion on [this PR](https://www.google.com/url?q=https://github.com/MetaMask/core/pull/3827%23discussion_r1469377179&sa=D&source=editors&ust=1716475694980436&usg=AOvVaw3oTcfVgfRuiQFSDChZPrFM))
```

Comment:
```suggestion
Unit tests understanding is a bit more diffuse. As there is no consensus on how they should be written and what they should test. But generically, [as captured by Martin Fowler](https://martinfowler.com/articles/2021-test-shapes.html), there's two lines of thought: solitary unit tests and sociable unit tests. Within MetaMask there are multiple examples from tests and discussions highlighting the two types (for example you can check the discussion on [this PR](https://github.com/MetaMask/core/pull/3827#discussion_r1469377179))
```

> As there is no consensus on how they should be written and what they should test.

This is a bold, sweeping statement, especially given the near-universal usage of unit testing in the industry. Could you provide more context on what you mean by this, and provide any reputable sources citing this lack of consensus?

> there's two lines of thought: solitary unit tests and sociable unit tests

Should we provide a summary of what these terms mean, how they are different, and why they are important? Are we expecting the reader of this doc to read through the entire linked blog article to find out?

Also, if these two terms are central to our understanding of unit testing, should they be discussed in the unit testing doc itself?

> `[this PR](https://www.google.com/url?q=https://github.com/MetaMask/core/pull/3827%23discussion_r1469377179&sa=D&source=editors&ust=1716475694980436&usg=AOvVaw3oTcfVgfRuiQFSDChZPrFM))`

There are quite a few comments in the linked thread. Should we provide an indication of which parts of the discussion we're trying to emphasize, and how they are relevant to this paragraph? Should the reader of this doc be expected to read through the entire discussion thread and discern the implications themselves?

Also, is there a reason we're using a google redirect link here?

Reply:
Ok this is a big comment, let me address each point:
> This is a bold, sweeping statement, especially given the near-universal usage of unit testing (and TDD) in the industry. Could you maybe provide more context on what you mean by this?

Universal usage does not mean universal understanding on how to test a unit. And that section for the unit tests is there to recognise that there are different types of unit tests. Being the most common ones what Martin Fowler names solitary unit tests and sociable unit tests. And there are multiple examples of both in our codebase. The PR discussion highlighted, shows the two lines of though being discussed (i.e. where are the limits of a unit test, where and what should we mock).

> Should we provide a summary of what these terms mean, how they are different, and why they are important? Are we expecting the reader of this doc to read through the linked blog article to find out?

Those could be added in the unit tests doc. I haven't added anything there, neither summarised the concept of sociable vs solitary unit tests as that was not the scope of this PR. The intended scope was to provide some generic guidelines on when to rely on each type of test (this is a follow up of adding the [UI integration tests into the Extension)](https://github.com/MetaMask/metamask-extension/pull/24428).

> Also, if these two terms are central to our understanding of unit testing, should they be discussed in the unit testing doc itself?

Yes. But maybe on a different PR where we expand our unit test documentation.

> Should the reader of this doc be expected to read through the entire discussion thread and discern the implications themselves?

I honestly think that it would be good to read the entire thread, so that reader can get a good sense on how different developers where advocating for different patterns without realising that the limits of a unit test are quite different for each.

> Also, is there a reason we're using a google redirect link here?

No.. Not sure how that turned into a google redirect link 🤦‍♂️ . Fixing it.